### PR TITLE
Raft API Client

### PR DESCRIPTION
**Pull request by SimonRichardson**
~~**Requires https://github.com/juju/juju/pull/13342 to land first**~~

The following adds a raftlease client. The aim is to provide a robust
way of sending requests whilst being honest about the error messages
that fail. In particular, we drop requested commands when we don't have a
connection and it's up to the callee to retry again. Previous
implementations just dropped them on the floor.

## QA steps

Current tests pass. As this hasn't been turned on yet, we just need
to ensure that existing implementations work.

```sh
$ juju bootstrap lxd test
$ juju model-config -m controller logging-config="<root>=INFO;juju.worker.lease.raft=TRACE;juju.core.raftlease=TRACE;juju.worker.globalclockupdater.raft=TRACE;juju.worker.raft=TRACE"
$ juju enable-ha
$ juju debug-log -m controller
```

### CODE REVIEW

File: api/raftlease/client.go [hunk 1]
```diff
@@ -0,0 +1,609 @@
...
+	}
+
+	// Kill off any servers that have fallen away. In a highly available setup,
+	// this might actually be really painful for us. We should really consider
+	// if this is wise or not? During flapping instances, we take a long time
+	// to connect and this could cause stalling of message sending.
+	//
+	// Instead we should probably just tombstone them first time around, then
+	// next time around we should reap them.
+	for id, remote := range c.servers {
```
Comment (achilleasa) [inline api/raftlease/client.go:422]: To handle flapping instances you could add these into a reap set (for this case a map[server]goneTimestamp should suffice) which gets periodically reaped in your main loop (i.e. kill anything where goneTimestamp < threshold). In the meantime, if you happen to see the server again just remove it from the reap set.

File: api/raftlease/client.go [hunk 2]
```diff
@@ -0,0 +1,609 @@
...
+		close(r.stopConnecting)
+		r.stopConnecting = nil
+	}
+	r.config.APIInfo.Addrs = []string{addr}
+}
+
+// Request performs a request against a specific api.
+func (r *remote) Request(ctx context.Context, command *raftlease.Command) error {
+	if r.client == nil {
+		r.config.Logger.Errorf("No client dropping command: %v", command)
```
Comment (achilleasa) [inline api/raftlease/client.go:504]:
```suggestion
		r.config.Logger.Errorf("No attached client instance; dropping command: %v", command)
```

File: api/raftlease/client.go [hunk 3]
```diff
@@ -0,0 +1,609 @@
...
+
+func (r *remote) loop() error {
+	var delay <-chan time.Time
+	for {
+		if r.api == nil {
+			if r.connect() {
+				delay = nil
+			} else {
+				// Apply a back-off if we where unable to connect.
+				delay = r.config.Clock.After(time.Second * 5)
```
Comment (achilleasa) [inline api/raftlease/client.go:546]: If you want something fancier for backoffs: https://github.com/juju/retry/blob/master/retry.go#L239
(Not sure if this one is pulled into the 2.9 branch though)

**Review by achilleasa — COMMENTED** [review]
LGTM; running QA

**Review by achilleasa — APPROVED** [review]
QA pass

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**